### PR TITLE
Feat/slepukhin/saved professionals

### PR DESCRIPTION
**Pull request by vitslepukhin**

### CODE REVIEW

File: src/app/api/services/professionals.service.ts [hunk 1]
```diff
@@ -595,7 +595,9 @@ module ProfessionalsService {
     /**
      * Multiple values may be separated by commas.
      */
-    pkIn?: number;
+
+    // TODO Fix swagger number -> string
```
Comment (enkryptor) [inline src/app/api/services/professionals.service.ts:599]: this TODO will be lost when the code is autogenerated

let's move the type casting to the consuming code

File: src/app/saved-professionals/components/user-saved-professionals-list/user-saved-professionals.component.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  public readonly isLoaded$: Observable<boolean>;
+
+  public savedProfessionals$: Observable<ProfessionalList[]>;
+
+  constructor(private readonly professionalsService: ProfessionalsService) {}
+
+  public ngOnInit(): void {
+    this.requestToLoadAllUserSavedProfessionals();
+    this.savedProfessionals$ = this.isLoaded$.pipe(
+      filter(isLoaded => isLoaded),
```
Comment (enkryptor) [inline src/app/saved-professionals/components/user-saved-professionals-list/user-saved-professionals.component.ts:30]: why we can't filter userSavedProfessionals$ itself?

File: src/app/store/current-user/saved-professionals/saved-professionals.state.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+  @Action(UserSavedProfessionalActions.LoadAllUserSavedProfessionals)
+  public loadAllUserSavedProfessionals({ patchState }: StateContext<UserSavedProfessionalStateModel>) {
+    return this.accountsService.accountsSavedProfessionalsList({}).pipe(
+      tap(({ results }) => {
+        const byId: EntityState<UserSavedProfessional>['byId'] = getByIdFromArray<UserSavedProfessional>(results);
+        patchState({
+          ids: uniqueArray([...results.map(({ id }) => id)]),
+          byId,
+          isLoaded: true,
```
Comment (enkryptor) [inline src/app/store/current-user/saved-professionals/saved-professionals.state.ts:43]: `isLoaded` is probably redundant:
ids is null → isLoaded: false;
ids is object → isLoaded: true;

File: src/app/store/current-user/saved-professionals/saved-professionals.state.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  defaults: EmptyEntityState,
+})
+export class UserSavedProfessionalState {
+  constructor(private readonly accountsService: AccountsService) {}
+
+  @Action(UserSavedProfessionalActions.LoadAllUserSavedProfessionals)
+  public loadAllUserSavedProfessionals({ patchState }: StateContext<UserSavedProfessionalStateModel>) {
+    return this.accountsService.accountsSavedProfessionalsList({}).pipe(
+      tap(({ results }) => {
+        const byId: EntityState<UserSavedProfessional>['byId'] = getByIdFromArray<UserSavedProfessional>(results);
```
Comment (enkryptor) [inline src/app/store/current-user/saved-professionals/saved-professionals.state.ts:39]: both byId and `ids` are calculated from `results`

we can calculate them in the selectors, leaving `results` in the state
selectors will cache the data, so performance wouldn't be affected

File: src/app/store/current-user/saved-professionals/saved-professionals.actions.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
+/* eslint-disable max-classes-per-file */
+
+import { Professional } from '@app/api/models';
+
+export class RequestToLoadAllUserSavedProfessionals {
+  public static readonly type = '[UserSavedProfessional] Request to load all user saved professionals';
+}
+
+export class LoadAllUserSavedProfessionals {
```
Comment (enkryptor) [inline src/app/store/current-user/saved-professionals/saved-professionals.actions.ts:9]: It's unclear, when the developer should use either the `RequestToLoadAllUserSavedProfessionals` or the  `LoadAllUserSavedProfessionals` action. This sounds like a leaked abstraction. Consider rethinking the interface.

File: src/app/api/services/professionals.service.ts (hunk 1 above)
Comment (vitslepukhin) [reply src/app/api/services/professionals.service.ts:599]: 1) The code using API is right - mistake is in API. I don't want to spoil right code. So let it stay here.
2) If after generating API nothing breaks - it is great. If it still breaks we need to pay attention to this and we will do it thanks Typescript.

File: src/app/store/current-user/saved-professionals/saved-professionals.state.ts (hunk 3 above)
Comment (vitslepukhin) [reply src/app/store/current-user/saved-professionals/saved-professionals.state.ts:43]: Probably yes but no. 
1) It is let us don't check values to null everywhere where state is used. As empty initial state is { ids: [], byId: {}, isLoading: false } not null.
2) It explicitly express state of data. Don't need to rely on null checks.

File: src/app/saved-professionals/components/user-saved-professionals-list/user-saved-professionals.component.ts (hunk 2 above)
Comment (vitslepukhin) [reply src/app/saved-professionals/components/user-saved-professionals-list/user-saved-professionals.component.ts:30]: Of course we can. In here I prefer explicitly check isLoading but not results===null or anything else like this.

File: src/app/store/current-user/saved-professionals/saved-professionals.state.ts (hunk 4 above)
Comment (vitslepukhin) [reply src/app/store/current-user/saved-professionals/saved-professionals.state.ts:39]: Yes we can do it in a selectors but we shouldn't as it is [normalized state](https://redux.js.org/recipes/structuring-reducers/normalizing-state-shape).

File: src/app/store/current-user/saved-professionals/saved-professionals.actions.ts (hunk 5 above)
Comment (vitslepukhin) [reply src/app/store/current-user/saved-professionals/saved-professionals.actions.ts:9]: Any proposals?

File: src/app/store/current-user/saved-professionals/saved-professionals.state.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
+import { Injectable } from '@angular/core';
+import { UserSavedProfessional } from '@app/api/models';
+import { AccountsService } from '@app/api/services';
+import { Action, State, StateContext } from '@ngxs/store';
+import { tap } from 'rxjs/operators';
+import * as UserSavedProfessionalActions from './saved-professionals.actions';
+
+export const EmptyUserSavedProfessionalState: UserSavedProfessional[] = null;
```
Comment (enkryptor) [inline src/app/store/current-user/saved-professionals/saved-professionals.state.ts:8]: Please use camelCase or UPPER_CASE for variable names.

**Review by enkryptor — APPROVED** [review]

**Pull request merged**